contributors/design-proposals/scheduling/scheduling-framework.md

```diff
@@ -1,436 +1,618 @@
+    PreFilter      []Plugin
+    Filter         []Plugin
+    PostFilter     []Plugin
+    Score          []Plugin
+    NormalizeScore []Plugin
+    Reserve        []Plugin
+    Permit         []Plugin
+    PreBind        []Plugin
+    Bind           []Plugin
+    PostBind       []Plugin
+    UnReserve      []Plugin
+}
+
+type Plugin struct {
+    Name   string
+    Weight int // Only valid for Score plugins
+}
+
+type PluginConfig struct {
+    Name string
+    Args json.RawMessage
+}
+```
+
+Example:
+
+```json
+{
+  "plugins": {
+    "preFilter": [
+      {
+        "name": "PluginA"
+      },
+      {
+        "name": "PluginB"
+      },
+      {
+        "name": "PluginC"
+      }
+    ],
+    "score": [
+      {
+        "name": "PluginA",
+        "weight": 30
+      },
+      {
+        "name": "PluginX"
+      },
+      {
+        "name": "PluginY"
+      }
+    ]
+  },
+  "pluginConfig": [
+    {
+      "name": "PluginX",
+      "args": {
+        "favorite_color": "#326CE5",
+        "favorite_number": 7,
+        "thanks_to": "thockin"
+      }
+    }
+  ]
+}
+```
+
+### Enable/Disable
+
+When specified, the list of plugins for a particular extension point are the
+only ones enabled. If an extension point is omitted from the config, then the
+default set of plugins is used for that extension point.
+
+### Change Evaluation Order
+
+When relevant, plugin evaluation order is specified by the order the plugins
+appear in the configuration. A plugin that registers for multiple extension
+points can have different ordering at each extension point.
+
+### Optional Args
+
+Plugins may receive arguments from their config with arbitrary structure.
+Because one plugin may appear in multiple extension points, the config is in a
+separate list of `PluginConfig`.
+
+For example,
+
+```json
+{
+   "name": "ServiceAffinity",
+   "args": {
+      "LabelName": "app",
+      "LabelValue": "mysql"
+   }
+}
+```
+
+```go
+func NewServiceAffinity(args json.RawMessage, h PluginHandle) (Plugin, error) {
+    var config struct {
+        LabelName, LabelValue string
+    }
+    if err := json.Unmarshal(args, &config); err != nil {
+        return nil, errors.Wrap(err, "could not parse args")
+    }
+    //...
+}
+```
+
+### Backward compatibility
+
+The current `KubeSchedulerConfiguration` kind has `apiVersion:
+kubescheduler.config.k8s.io/v1alpha1`. This new config format will be either
+`v1alpha2` or `v1beta1`. When a newer version of the scheduler parses a
+`v1alpha1`, the "policy" section will be used to construct an equivalent plugin
+configuration.
+
+*Note: Moving `KubeSchedulerConfiguration` to `v1` is outside the scope of this
+design, but see also
+https://github.com/kubernetes/enhancements/blob/master/keps/sig-cluster-lifecycle/0032-create-a-k8s-io-component-repo.md
+and https://github.com/kubernetes/community/pull/3008*
+
+# USE CASES
+
+These are just a few examples of how the scheduling framework can be used.
+
+## Batch Scheduling
```

Comment:
We should call it Gang Scheduling or "Coscheduling". Batch Scheduling does not necessarily require a group of pods to be scheduled together.